metaflow/runtime.py

```diff
@@ -160,13 +163,24 @@ def __init__(
             for deco in step.decorators:
                 deco.runtime_init(flow, graph, package, self._run_id)
 
+    def _update_rerun_steps(self):
+        """
+        Any steps following steps to be rerun should also be included as
+        "rerun" steps. This is to ensure that the flow is executed correctly.
+        """
+        for step_name in self._graph.sorted_nodes:
+            if step_name in self._rerun_steps:
+                out_funcs = self._graph[step_name].out_funcs or []
+                for next_step in out_funcs:
+                    self._rerun_steps.add(next_step)
+
     def _new_task(self, step, input_paths=None, **kwargs):
         if input_paths is None:
             may_clone = True
         else:
             may_clone = all(self._is_cloned[path] for path in input_paths)
 
-        if step in self._clone_steps:
+        if step in self._rerun_steps:
             may_clone = False
```

Comment:
In general, I think with this new way, we can clean out all the may_clone, etc flags taht are being set. I think it may be confusing because I don't think they are used anymore so that would simplify the code a bit and make it clearer that all the resume logic is now in one place (the clone functions) as opposed to scattered around the runtime code.

Reply:
I agree, but the code refactor should be in a different PR (I want this fix PR to be concise).

Reply:
OK. Let's do a separate PR to clean that up then.